proxy/src/main/java/net/md_5/bungee/UserConnection.java

```diff
@@ -374,17 +375,22 @@ public void operationComplete(ChannelFuture future) throws Exception
                     future.channel().close();
                     pendingConnects.remove( target );
 
-                    ServerInfo def = updateAndGetNextServer( target );
-                    if ( request.isRetry() && def != null && ( getServer() == null || def != getServer().getInfo() ) )
+                    ServerConnection server = getServer();
+                    ServerConnectFailEvent event = new ServerConnectFailEvent( UserConnection.this, server, request );
+                    if ( !bungee.getPluginManager().callEvent( event ).isCancelled() )
```

Comment:
If you cancel this doesn't the player just get stuck in limbo

Reply:
Yes, when your plugin cancels the event without disconnect or sendMessage,
- If a player tries to log in, "Joining World..." will be displayed and stuck until timeout.
- If the player tries to move with `/server`, etc., the server move will be canceled and nothing will happen.

I think this is the correct behavior, since the point of canceling this event is to handle it on the plugin side.

Your plugin can for example do the following
- Start the server asynchronously while the player is connected and stuck, then connect the player after the server starts. You can kick the server if it fails to start.

Do you have any ideas for a more appropriate behavior for cancellations?
I will be glad to implement it if you request it!

Reply:
Let me think about it. I think this is a hard one because already have callback and connect events